jmprathab/MyHome
# Added Payments Scheduling features

## 🚀 Description

Added features so a community admin can schedule a payment for a house member

I added new controllers and endpoints for this. These endpoints are as follows:

- /members/payments (Create a payment)

- /members/payments/{paymentId} (Show details about a payment)

- /members/{memberId}/payments (Show the payments for the specified member)

- /communities/{communityId}/admins/{adminId}/payments (view the payments for a specified community admin that they scheduled)

This required the addition of a set of Payments to the domain.House class for storing these payments

## 📄 Motivation and Context

This has been implemented as the issue #48 requested that this feature should be added

## 🧪 How Has This Been Tested?

I did development and testing on Ubuntu 20.04
I used Postman to test my features with the following steps:

1. Start the MyHomeApplication using the jar file/docker

2. Open Postman

3. Run the Create User and Login requests. From the login request header, record the token (use this bearer token for all following requests 

4. Run the List all Users request and record 2 user Ids, one to be added to a house as a member and one to be used as an admin

5. Run the Show Communities request and record one communityId

6. Run the List all houses request and record the houseId

7. Run the add member to house request with the houseId from 6 and in the body the userId and name of the user from 4 and record the produced memberId

8. Run the add admin to community request with the community id got from 5 and admin id in body got from step 4

9. Then open a new post request with the following endpoint: http://localhost:8080/members/payments and in the body choose JSON format and enter the following format with date in format yyyy-MM-dd
`{
  "type": "string",
  "description": "string",
  "recurring": true,
  "charge": 0,
  "dueDate": "string",
  "adminId": "string",
  "memberId": "string"
}`

Record the produced paymentId

10. Hit the endpoint /members/payments/{paymentId} with the id obtained from step 9 and verify that the details produced match 
11.Hit the endpoint /members/{memberId}/payments with memberId from 7 and verify the payment shows up
12. Hit the endpoint /communities/{communityId}/admins/{adminId}/payments with the communityId from step 5 and adminId from step 4 and verify the payment shows up

13. Add another different payment and repeat 10, 11 and 12 to verify they all show up

## 📷 Screenshots (if appropriate)

[Schedule Payment Response](https://ibb.co/RPCwcH5) (Apologies for 16 second response time, was running in debug and forgot a breakpoint was hit!
[Payment Info Response](https://ibb.co/7ySFDyV)
[Member payments response](https://ibb.co/qgMNbxH)
[Community admin scheduled payments response](https://ibb.co/t4rDvbx)

## 📦 Types of changes




## ✅ Checklist




- [X] My code follows the code style of this project(Do your best to follow code styles. If none apply just skip this).
